contracts/test/LightClientUpgrade.t.sol

```diff
@@ -0,0 +1,146 @@
+
+    LCV1.LightClientState public stateV1;
+
+    address public admin;
+    address public proxy;
+
+    // deploy the first implementation with its proxy
+    function setUp() public {
+        (proxy, admin, stateV1) = deployer.run();
+        lcV1Proxy = LCV1(proxy);
+    }
+
+    function testCorrectInitialization() public {
+        assert(lcV1Proxy.blocksPerEpoch() == 10);
+        assert(lcV1Proxy.currentEpoch() == 0);
+
+        {
+            (
+                uint64 viewNum,
+                uint64 blockHeight,
+                BN254.ScalarField blockCommRoot,
+                BN254.ScalarField feeLedgerComm,
+                BN254.ScalarField stakeTableBlsKeyComm,
+                BN254.ScalarField stakeTableSchnorrKeyComm,
+                BN254.ScalarField stakeTableAmountComm,
+                uint256 threshold
+            ) = lcV1Proxy.genesisState();
+            LCV1.LightClientState memory expectedState = LCV1.LightClientState(
+                viewNum,
+                blockHeight,
+                blockCommRoot,
+                feeLedgerComm,
+                stakeTableBlsKeyComm,
+                stakeTableSchnorrKeyComm,
+                stakeTableAmountComm,
+                threshold
+            );
+
+            assertEq(abi.encode(expectedState), abi.encode(stateV1));
+        }
+        {
+            (
+                uint64 viewNum,
+                uint64 blockHeight,
+                BN254.ScalarField blockCommRoot,
+                BN254.ScalarField feeLedgerComm,
+                BN254.ScalarField stakeTableBlsKeyComm,
+                BN254.ScalarField stakeTableSchnorrKeyComm,
+                BN254.ScalarField stakeTableAmountComm,
+                uint256 threshold
+            ) = lcV1Proxy.finalizedState();
+            LCV1.LightClientState memory expectedState = LCV1.LightClientState(
+                viewNum,
+                blockHeight,
+                blockCommRoot,
+                feeLedgerComm,
+                stakeTableBlsKeyComm,
+                stakeTableSchnorrKeyComm,
+                stakeTableAmountComm,
+                threshold
+            );
+
+            assertEq(abi.encode(expectedState), abi.encode(stateV1));
+        }
+
+        bytes32 stakeTableComm = lcV1Proxy.computeStakeTableComm(stateV1);
+        assertEq(lcV1Proxy.votingStakeTableCommitment(), stakeTableComm);
+        assertEq(lcV1Proxy.frozenStakeTableCommitment(), stakeTableComm);
+        assertEq(lcV1Proxy.votingThreshold(), stateV1.threshold);
+        assertEq(lcV1Proxy.frozenThreshold(), stateV1.threshold);
+    }
+
+    // that the data remains the same after upgrading the implementation
+    function testUpgradeSameData() public {
+        // Upgrade LightClient and check that the genesis state is not changed and that the new
+        // field
+        // of the upgraded contract is set to 0
+        lcV2Proxy = LCV2(upgrader.run(admin, proxy));
+
+        assertEq(lcV2Proxy.newField(), 0);
```

Comment:
Yes, however I tried that and it did not work. I think the reason is that genesisState and finalizedState are at the beginning of storage. So it seems we have too options in order to be able to extend LightClientState struct with new fields:

* Add new (and upgraded) `LightClientState` fields to the contract  and provide some conversion method for reading old states.
* Put the `LightClientState` variable (`genesis` and `finalizedState`) into a map (looks weird but seems to be the cleanest/most flexible approach).

WDYT?

Reply:
Ok that makes sense - the mapping approach can look weird but sounds cleaner for sure. From reading, using a storage gap would lead to higher gas costs too so I think the mapping approach may be best based on what we know right now.

Reply:
Done in 0118f89e7f7f9768cd68450ceb4b7c943e76ffe7.